### ISSUE TITLE
Chunked mean/stddev is failing in some cases

### ISSUE THREAD

**Issue by mexxexx** [TDAmeritrade/stumpy#198]
I noticed that in some cases the chunked mean/stddev computation is failing, with the following error: 

```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-9-4f878a46a7c1> in <module>
      9 
     10 start = time.time()
---> 11 P = stumpy.stump(T, m)
     12 print(time.time() - start)

~/Dokumente/Programmierung/stumpy/stumpy/stump.py in stump(T_A, m, T_B, ignore_trivial)
    422     stop = l
    423 
--> 424     all_start_profiles, indices[start, :] = _get_first_stump_profile(
    425         start, T_A, T_B, m, excl_zone, M_T, Σ_T, μ_Q, σ_Q, ignore_trivial
    426     )

~/Dokumente/Programmierung/stumpy/stumpy/stump.py in _get_first_stump_profile(start, T_A, T_B, m, excl_zone, M_T, Σ_T, μ_Q, σ_Q, ignore_trivial)
     82     else:
     83         if ignore_trivial:
---> 84             P, I = stamp._mass_PI(
     85                 T_B[start : start + m], T_A, M_T, Σ_T, start, excl_zone
     86             )

~/Dokumente/Programmierung/stumpy/stumpy/stamp.py in _mass_PI(Q, T, M_T, Σ_T, trivial_idx, excl_zone, left, right)
     48         Matrix profile indices
     49     """
---> 50     D = core.mass(Q, T, M_T, Σ_T)
     51 
     52     if trivial_idx is not None:

~/Dokumente/Programmierung/stumpy/stumpy/core.py in mass(Q, T, M_T, Σ_T)
    653 
    654         QT = sliding_dot_product(Q, T)
--> 655         μ_Q, σ_Q = compute_mean_std(Q, m)
    656         μ_Q = μ_Q[0]
    657         σ_Q = σ_Q[0]

~/Dokumente/Programmierung/stumpy/stumpy/core.py in compute_mean_std(T, m)
    313                 stop = min(start + chunk_size + m - 1, T.shape[-1])
    314 
--> 315                 tmp_mean = np.mean(rolling_window(T[start:stop], m), axis=T.ndim)
    316                 mean_chunks.append(tmp_mean)
    317                 tmp_std = np.nanstd(rolling_window(T[start:stop], m), axis=T.ndim)

~/Dokumente/Programmierung/stumpy/stumpy/core.py in rolling_window(a, window)
     51     strides = a.strides + (a.strides[-1],)
     52 
---> 53     return np.lib.stride_tricks.as_strided(a, shape=shape, strides=strides)
     54 
     55 

~/miniconda3/envs/linac3/lib/python3.8/site-packages/numpy/lib/stride_tricks.py in as_strided(x, shape, strides, subok, writeable)
    101         interface['strides'] = tuple(strides)
    102 
--> 103     array = np.asarray(DummyArray(interface, base=x))
    104     # The route via `__interface__` does not preserve structured
    105     # dtypes. Since dtype should remain unchanged, we set it explicitly.

~/miniconda3/envs/linac3/lib/python3.8/site-packages/numpy/core/_asarray.py in asarray(a, dtype, order)
     83 
     84     """
---> 85     return array(a, dtype, copy=False, order=order)
     86 
     87 

ValueError: negative dimensions are not allowed
```

A test to reproduce it:
```
Q, T = np.random.uniform(-1000, 1000, [8]), np.random.uniform(-1000, 1000, [64])

def test_compute_mean_std_chunked(Q, T):
    config.STUMPY_MEAN_STD_NUM_CHUNKS = 2

    m = Q.shape[0]

    left_μ_Q, left_σ_Q = naive_compute_mean_std(Q, m)
    left_M_T, left_Σ_T = naive_compute_mean_std(T, m)
    right_M_T, right_Σ_T = core.compute_mean_std(T, m)
    right_μ_Q, right_σ_Q = core.compute_mean_std(Q, m)

    npt.assert_almost_equal(left_μ_Q, right_μ_Q)
    npt.assert_almost_equal(left_σ_Q, right_σ_Q)
    npt.assert_almost_equal(left_M_T, right_M_T)
    npt.assert_almost_equal(left_Σ_T, right_Σ_T)

test_compute_mean_std_chunked(Q, T)
```

Computing mean and stddev of `T` works, but it fails when computing the mean/stddev of `Q`.

I think the error happens here if `stop - start < m` (in `core.compute_mean_std_chunked`):
``` 
for chunk in range(num_chunks):
    start = chunk * chunk_size
    stop = min(start + chunk_size + m - 1, T.shape[-1])

    tmp_mean = np.mean(rolling_window(T[start:stop], m), axis=T.ndim)
```
In that case the loop should stop, because it is the last window which should be empty, however `rolling_window` produces an error if `len(a) < window`.

Adding a test against it seems to help (Tests pass):
``` 
for chunk in range(num_chunks):
     start = chunk * chunk_size
     stop = min(start + chunk_size + m - 1, T.shape[-1])
     if stop - start < m:
         break
     tmp_mean = np.mean(rolling_window(T[start:stop], m), axis=T.ndim)
```

**Comment by seanlaw** [comment]
Can you please add your example to the unit test quite? That's a good one

I think another potential issue is, say, when all of your chunks have a `chunk_size` that are smaller than `m`. That can happen if you, say, choose `config.STUMPY_MEAN_STD_NUM_CHUNKS = T.shape[0]` (for a 1D example). Then, none of the means/stddevs are computed.

**Comment by mexxexx** [comment]
That's true!

Doing the same thing in the multidimensional case shows that there are some slicing errors. This fixes them:

```
tmp_mean = np.mean(rolling_window(T[start:stop], m), axis=T.ndim)
```
Should be 
```
tmp_mean = np.mean(rolling_window(T[..., start:stop], m), axis=T.ndim)
```

**Comment by seanlaw** [comment]
> Doing the same thing in the multidimensional case shows that there are some slicing errors. This fixes them:

> tmp_mean = np.mean(rolling_window(T[start:stop], m), axis=T.ndim)
>Should be

>tmp_mean = np.mean(rolling_window(T[..., start:stop], m), axis=T.ndim)

Great catch. How was it possible that `tests_core.py::test_compute_mean_std_multidimensional(Q, T)` was passing?

**Comment by seanlaw** [comment]
> How was it possible that tests_core.py::test_compute_mean_std_multidimensional(Q, T) was passing?

Ohh, I get it. It was selecting the first `[start:stop]` rows (or dimensions) from `T` and then computing the mean/stddev. So, technically, the answer is still correct but you run the risk of having very long rows. Versus what we were trying to achieve here was to break up the length of the time series into chunks and not the dimensions.

**Comment by mexxexx** [comment]
Exactly, and if you have less then `m` dimensions, then you would again end up with empty chunks.

**Comment by seanlaw** [comment]
This is kinda nasty. Let me know if you need any assistance resolving this. I'd like to get a resolution in for this ASAP so I can include it in the next release (I'm hoping this weekend). Where `scrump` would be the focus.

**Comment by mexxexx** [comment]
I have a PR almost ready

**Comment by seanlaw** [comment]
For safety, does it make any sense to write a test that increments the number of chunks? Something like:

```
Q, T = np.random.uniform(-1000, 1000, [8]), np.random.uniform(-1000, 1000, [64])

def test_compute_mean_std_chunked(Q, T):
    left_μ_Q, left_σ_Q = naive_compute_mean_std(Q, m)
    left_M_T, left_Σ_T = naive_compute_mean_std(T, m)

    for chunks in range(Q.shape[0]):
        config.STUMPY_MEAN_STD_NUM_CHUNKS = chunks

        m = Q.shape[0]

        right_M_T, right_Σ_T = core.compute_mean_std(T, m)
        right_μ_Q, right_σ_Q = core.compute_mean_std(Q, m)

        npt.assert_almost_equal(left_μ_Q, right_μ_Q)
        npt.assert_almost_equal(left_σ_Q, right_σ_Q)
        npt.assert_almost_equal(left_M_T, right_M_T)
        npt.assert_almost_equal(left_Σ_T, right_Σ_T)

    config.STUMPY_MEAN_STD_NUM_CHUNKS = 1
```

Notice the last line where we purposely reset it back to `1` for all other tests.

**Comment by mexxexx** [comment]
Personally, I dont think it's necessary, but I can add them. I added two tests for the one and multidimensional case, once with `config.STUMPY_MEAN_STD_NUM_CHUNKS=2` and once with `config.STUMPY_MEAN_STD_NUM_CHUNKS=128`. The second case covers he cases where `chunk_size<m`.

**Comment by seanlaw** [comment]
That's fair. Just please remember to reset to `1`. Otherwise, stick to what you are doing!

**Comment by seanlaw** [comment]
Fixed in #199
